doc/decisions/1_in_discussion/reference_counting.md

```diff
@@ -0,0 +1,286 @@
+        ref->free(ref);
+}
+```
+
+This structure is not meant to be allocated separately, instead it should be embedded like this:
+
+```c
+typedef struct Key {
+    // [...] other stuff
+    struct ref refcount;
+} Key;
+```
+
+Now to create a `Key`, we need a function that takes a `const struct ref *` and frees the `Key` that we can put into `struct ref`.
+
+```c
+// Also used in Linux Kernel
+// See: https://radek.io/2012/11/10/magical-container_of-macro/
+#define container_of(ptr, type, member) ((type *)((char *)(ptr) - offsetof(type, member)))
+
+static void keyFree(const struct ref *ref)
+{
+    Key * key = container_of(ref, struct Key, refcount);
+    // [...] other cleanup
+    free(key);
+}
+```
+
+This function would only be private.
+The public API would use functions like this:
+
+```c
+Key * keyCreate() {
+  Key * key = malloc(sizeof(struct Key));
+  // [...] other initialization
+  key->refcount = (struct ref){keyFree, 1};
+  return key;
+}
+
+void keyRetain(Key * key) {
+  ref_inc(key->refcount);
+}
+
+void keyRelease(Key * key) {
+  ref_dec(key->refcount);
+}
+```
+
+These functions are needed to keep `struct Key` opaque.
+
+A big advantage here would be that things like `mmapstorage` that do not use normal heap allocations, could set their own `free` pointer in `struct ref`.
+
+### Reference counting in allocator
+
+While we cannot have automatic reference counting (ARC), we could implement most parts of ARC other than the automatic counting inserted by a compiler.
+In an ARC system, every heap memory allocation has a reference count, which is often stored as a hidden header.
+Essentially, instead of calling `malloc(data_size)` the system does `malloc(header_size+data_size)` and uses the additional `header_size` bytes to store data like the reference count.
+Crucially, the pointer returned to the caller still points after the header data, which is therefore hidden.
+
+Since (almost) all allocations for Elektra's data structures go through `elektraMalloc`, we could implement something similar.
+In `elektraMalloc` we allocate a few extra bytes (for alignment reasons this would be `sizeof(size_t)`) and use them for reference counting.
+The reference counting would be done via generic `elektraRetain` and `elektraRelease` functions.
+These functions increment and decrement the counter in the hidden header.
+Additionally, `elektraRelease` would also call `free` when the count reaches zero.
+
+Like in the "Retain/Release/Free" option above, the hidden header data could also be used to store a pointer to a `free` function.
+Then things like `mmapstorage` could be covered as well.
+
+```c
+// Based on: https://codereview.stackexchange.com/questions/146561/reference-counting-in-c99
+
+// Note: should NOT call free() on pointer
+typedef void (*Destructor)(const void*);
+
+typedef struct {
+  Destructor destroy;
+  unsigned int count;
+} ObjectRecord;
+
+static ObjectRecord *toRecord(void *object) { return (ObjectRecord*)(object) - 1; }
```

Comment:
```suggestion
static ObjectRecord *objectToRecord(void *object) { return (ObjectRecord*)(object) - 1; }
```

Reply:
Wait ... is this really `object - 1`? Shouldn't it be `object - sizeof(ObjectRecord)`?

Reply:
Yes, it works because of the parentheses. `(ObjectRecord*)(object - 1)` would be wrong, but by casting first and then subtracting the pointer arithmetic is done in terms of `sizeof(ObjectRecord)`. One of the weird and wonderful properties of C. See also the first answer of the StackExchange link a few lines above.